File: pi/lib/herdr_workflow/tiering.py
"""Pure verification-tiering logic. Git diff text is passed in, never fetched here."""
import re
from pathlib import Path

VERIFIER_ROLES = ("security-verifier", "agents-verifier", "quality-verifier", "performance-verifier", "openspec-verifier")
TEST_VERIFIER = "test-verifier"


def review_tier(diff_numstat, paths):
    lines = sum(sum(map(int, row.split("\t")[:2])) for row in diff_numstat.splitlines() if row and "-" not in row)
    sensitive = any(re.search(r"(^|/)(auth|security|crypto|secret|permission|migration)(/|$)", path, re.I) for path in paths)
    docs_only = bool(paths) and all(re.search(r"(^|/)(README|AGENTS|CLAUDE|docs?/)|\.md$", path, re.I) for path in paths)
    if sensitive or len(paths) > 50 or lines > 100:
        return "full", VERIFIER_ROLES
    if docs_only and lines <= 10:
        return "trivial", ("quality-verifier", "openspec-verifier")
    return "lite", ("security-verifier", "agents-verifier", "quality-verifier", "openspec-verifier")


def eligible_verifier_roles(files):
    joined = "\n".join(files)
    eligible = set()
    if any(not path.endswith(".md") for path in files):
        eligible.add("quality-verifier")
    if re.search(r"(^|/)(auth|security|crypto|secret|permission|migration)(/|$)", joined, re.I):
        eligible.add("security-verifier")
    if any(Path(path).name in {"AGENTS.md", "CLAUDE.md"} for path in files):
        eligible.add("agents-verifier")
    if any(path.startswith("openspec/") or path.endswith(("openapi.yaml", "openapi.yml")) for path in files):
        eligible.add("openspec-verifier")
    if re.search(r"(performance|benchmark|cache|stream|batch|query|algorithm)", joined, re.I):
        eligible.add("performance-verifier")
    return sorted(eligible)


def file_manifest(diff_numstat, diff_text, files):
    stats = {row.split("\t", 2)[2]: {"added": row.split("\t", 2)[0], "removed": row.split("\t", 2)[1]} for row in diff_numstat.splitlines() if row and "\t" in row}
    hunks, current = {}, None
    for line in diff_text.splitlines():
        if line.startswith("+++ b/"):
            current = line[6:]
        elif current and line.startswith("@@"):
            hunks.setdefault(current, []).append(line)
    return [{"path": path, **stats.get(path, {"added": "?", "removed": "?"}), "hunks": [{"id": index + 1, "header": header} for index, header in enumerate(hunks.get(path, [])[:8])]} for path in files]


def applicable_instructions(root, files):
    """Local AGENTS.md/CLAUDE.md discovery — plain filesystem walk, no git involved."""
    found = set()
    for file in files:
        directory = (root / file).parent
        while directory >= root:
            for name in ("AGENTS.md", "CLAUDE.md"):
                candidate = directory / name
                if candidate.is_file():
                    found.add(str(candidate.relative_to(root)))
            if directory == root:
                break
            directory = directory.parent
    return sorted(found)

File: agentic-coding/src/workflow/tiering.ts
// Pure verification-tiering logic. Git diff text is passed in, never fetched here.
import fs from 'node:fs';
import path from 'node:path';

export const VERIFIER_ROLES = [
  'security-verifier',
  'agents-verifier',
  'quality-verifier',
  'usability-verifier',
  'performance-verifier',
  'openspec-verifier',
] as const;
export const TEST_VERIFIER = 'test-verifier';

const SENSITIVE_RE = /(^|\/)(auth|security|crypto|secret|permission|migration)(\/|$)/i;
const DOCS_RE = /(^|\/)(README|AGENTS|CLAUDE|docs?\/)|\.md$/i;

export function reviewTier(diffNumstat: string, paths: string[]): [string, readonly string[]] {
  const lines = diffNumstat
    .split('\n')
    .filter(row => row && !row.includes('-'))
    .reduce((sum, row) => sum + row.split('\t').slice(0, 2).reduce((a, b) => a + Number(b), 0), 0);
  const sensitive = paths.some(p => SENSITIVE_RE.test(p));
  const docsOnly = paths.length > 0 && paths.every(p => DOCS_RE.test(p));
  if (sensitive || paths.length > 50 || lines > 100) return ['full', VERIFIER_ROLES];
  if (docsOnly && lines <= 10) return ['trivial', ['quality-verifier', 'openspec-verifier']];
  return ['lite', ['security-verifier', 'agents-verifier', 'quality-verifier', 'openspec-verifier']];
}

const UI_SUFFIXES = new Set(['.css', '.scss', '.less', '.html', '.htm', '.jsx', '.tsx', '.vue', '.svelte', '.svg', '.png', '.gif', '.ico', '.woff', '.woff2', '.ttf', '.eot']);

export function eligibleVerifierRoles(files: string[]): string[] {
  const joined = files.join('\n');
  const eligible = new Set<string>();
  if (files.some(p => !p.endsWith('.md'))) eligible.add('quality-verifier');
  if (SENSITIVE_RE.test(joined)) eligible.add('security-verifier');
  if (files.some(p => ['AGENTS.md', 'CLAUDE.md'].includes(path.basename(p)))) eligible.add('agents-verifier');
  if (files.some(p => p.startsWith('openspec/') || p.endsWith('openapi.yaml') || p.endsWith('openapi.yml'))) eligible.add('openspec-verifier');
  if (
    files.some(p => {
      const ext = path.extname(p).toLowerCase();
      const parts = p.split('/').slice(0, -1);
      const name = path.basename(p);
      return UI_SUFFIXES.has(ext) || parts.some(part => ['frontend', 'ui', 'app'].includes(part)) || /^(theme|colors|tokens)\./i.test(name);
    })
  ) {
    eligible.add('usability-verifier');
  }
  if (/(performance|benchmark|cache|stream|batch|query|algorithm)/i.test(joined)) eligible.add('performance-verifier');
  return [...eligible].sort();
}

export interface FileManifestEntry {
  path: string;
  added: string;
  removed: string;
  hunks: Array<{ id: number; header: string }>;
}

export function fileManifest(diffNumstat: string, diffText: string, files: string[]): FileManifestEntry[] {
  const stats = new Map<string, { added: string; removed: string }>();
  for (const row of diffNumstat.split('\n')) {
    if (!row || !row.includes('\t')) continue;
    const [added, removed, ...rest] = row.split('\t');
    stats.set(rest.join('\t'), { added, removed });
  }
  const hunks = new Map<string, string[]>();
  let current: string | null = null;
  for (const line of diffText.split('\n')) {
    if (line.startsWith('+++ b/')) {
      current = line.slice(6);
    } else if (current && line.startsWith('@@')) {
      if (!hunks.has(current)) hunks.set(current, []);
      hunks.get(current)!.push(line);
    }
  }
  return files.map(file => ({
    path: file,
    ...(stats.get(file) ?? { added: '?', removed: '?' }),
    hunks: (hunks.get(file) ?? []).slice(0, 8).map((header, index) => ({ id: index + 1, header })),
  }));
}

/** Local AGENTS.md/CLAUDE.md discovery — plain filesystem walk, no git involved. */
export function applicableInstructions(root: string, files: string[]): string[] {
  const found = new Set<string>();
  for (const file of files) {
    let directory = path.dirname(path.join(root, file));
    while (directory === root || directory.startsWith(root + path.sep)) {
      for (const name of ['AGENTS.md', 'CLAUDE.md']) {
        const candidate = path.join(directory, name);
        if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
          found.add(path.relative(root, candidate));
        }
      }
      if (directory === root) break;
      directory = path.dirname(directory);
    }
  }
  return [...found].sort();
}
